tests/unit/s2n_server_supported_versions_extension_test.c

```diff
@@ -0,0 +1,178 @@
+
+    	/* Check that type and size are correct */
+    	uint16_t extension_type;
+    	s2n_stuffer_read_uint16(&extension, &extension_type);
+    	EXPECT_EQUAL(extension_type, TLS_EXTENSION_SUPPORTED_VERSIONS);
+    	uint16_t extension_length;
+    	s2n_stuffer_read_uint16(&extension, &extension_length);
+    	EXPECT_EQUAL(extension_length, s2n_stuffer_data_available(&extension));
+
+    	/* Check that the client can process the version */
+    	struct s2n_connection *client_conn;
+    	EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+    	EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+    	EXPECT_SUCCESS(s2n_extensions_server_supported_versions_recv(client_conn, &extension));
+    	EXPECT_EQUAL(client_conn->client_protocol_version, latest_version);
+    	EXPECT_EQUAL(client_conn->server_protocol_version, latest_version);
+    	EXPECT_EQUAL(client_conn->actual_protocol_version, latest_version);
+
+    	/* Clean up */
+    	EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    	EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    	EXPECT_SUCCESS(s2n_stuffer_free(&extension));
+
+    }
+
+    /* Client alerts if supported_version less than min supported by client */
+    {
+    	struct s2n_connection *client_conn;
+    	EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+    	EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+    	uint8_t unsupported_version_unknown = S2N_UNKNOWN_PROTOCOL_VERSION;
+
+    	uint16_t supported_version_length = 6;
+
+    	struct s2n_stuffer extension;
+    	s2n_stuffer_alloc(&extension, supported_version_length);
+
+    	EXPECT_SUCCESS(write_test_supported_version(&extension, unsupported_version_unknown));
+    	EXPECT_FAILURE_WITH_ERRNO(s2n_extensions_server_supported_versions_recv(client_conn, &extension), S2N_ERR_BAD_MESSAGE);
+    	EXPECT_EQUAL(get_alert(client_conn), ILLEGAL_PARAMETER_ALERT);
+
+    	EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    	EXPECT_SUCCESS(s2n_stuffer_free(&extension));
+
+    }
+
+    /* Client alerts if supported_version greater than max supported by client */
+    {
+    	struct s2n_connection *client_conn;
+    	EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+    	EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+    	uint8_t unsupported_version_gt_tls13 = 255;
+
+    	uint16_t supported_version_length = 6;
+
+    	struct s2n_stuffer extension;
+    	s2n_stuffer_alloc(&extension, supported_version_length);
+
+    	EXPECT_SUCCESS(write_test_supported_version(&extension, unsupported_version_gt_tls13));
+    	EXPECT_FAILURE_WITH_ERRNO(s2n_extensions_server_supported_versions_recv(client_conn, &extension), S2N_ERR_BAD_MESSAGE);
+    	EXPECT_EQUAL(get_alert(client_conn), ILLEGAL_PARAMETER_ALERT);
+
+    	EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    	EXPECT_SUCCESS(s2n_stuffer_free(&extension));
+
+    }
+
+    /* Client alerts if supported_version is empty */
```

Comment:
could you also add tests where client shouldn't alert when `S2N_ENABLE_TLS13_FOR_TESTING` is not set?

Reply:
will do!

Reply:
with the new gates in place for tls 1.3, this case will not actually be reached. instead we test that the gates work so that the extension won't be parsed at all if it's an earlier version: https://github.com/awslabs/s2n/blob/master/tests/unit/s2n_tls13_support_test.c